packages/flutter_tools/lib/src/version.dart

```diff
@@ -837,3 +784,122 @@ enum VersionCheckResult {
   /// A newer version is available.
   newVersionAvailable,
 }
+
+@visibleForTesting
+Future<void> checkVersionFreshness(FlutterVersion version, {
+  required DateTime localFrameworkCommitDate,
+  required DateTime? latestFlutterCommitDate,
+  required SystemClock clock,
+  required Cache cache,
+  required Logger logger,
+  Duration pauseTime = Duration.zero,
+}) async {
+  // Don't perform update checks if we're not on an official channel.
+  if (!kOfficialChannels.contains(version.channel)) {
+    return;
+  }
+
+  final Duration frameworkAge = clock.now().difference(localFrameworkCommitDate);
+  final bool installationSeemsOutdated = frameworkAge > versionAgeConsideredUpToDate(version.channel);
+
+  // Get whether there's a newer version on the remote. This only goes
+  // to the server if we haven't checked recently so won't happen on every
+  // command.
+  final VersionCheckResult remoteVersionStatus = latestFlutterCommitDate == null
+      ? VersionCheckResult.unknown
+      : latestFlutterCommitDate.isAfter(localFrameworkCommitDate)
+        ? VersionCheckResult.newVersionAvailable
+        : VersionCheckResult.versionIsCurrent;
+
+  // Do not load the stamp before the above server check as it may modify the stamp file.
+  final VersionCheckStamp stamp = await VersionCheckStamp.load(cache, logger);
+  final DateTime lastTimeWarningWasPrinted = stamp.lastTimeWarningWasPrinted ?? clock.ago(maxTimeSinceLastWarning * 2);
+  final bool beenAWhileSinceWarningWasPrinted = clock.now().difference(lastTimeWarningWasPrinted) > maxTimeSinceLastWarning;
+
+  // We show a warning if either we know there is a new remote version, or we couldn't tell but the local
+  // version is outdated.
+  final bool canShowWarning =
+    remoteVersionStatus == VersionCheckResult.newVersionAvailable ||
+      (remoteVersionStatus == VersionCheckResult.unknown &&
+        installationSeemsOutdated);
+
+  if (beenAWhileSinceWarningWasPrinted && canShowWarning) {
+    final String updateMessage =
+      remoteVersionStatus == VersionCheckResult.newVersionAvailable
+        ? newVersionAvailableMessage()
+        : versionOutOfDateMessage(frameworkAge);
+    logger.printStatus(updateMessage, emphasis: true);
+    await Future.wait<void>(<Future<void>>[
+      stamp.store(
+        newTimeWarningWasPrinted: clock.now(),
+        cache: cache,
+      ),
+      Future<void>.delayed(pauseTime),
+    ]);
+  }
+}
+
+/// The amount of time we wait before pinging the server to check for the
+/// availability of a newer version of Flutter.
+@visibleForTesting
+const Duration checkAgeConsideredUpToDate = Duration(days: 3);
```

Comment:
Can you remove `Version.checkAgeConsideredUpToDate`, `Version.versionAgeConsideredUpToDate` etc?  Having two copies to keep in sync seems unnecessary.

Reply:
I thought I did, my bad